## World Bank Projects (JSON exercise)
****
**Goal:** Using the world bank projects dataset (in file 'data/world_bank_projects.json'), provide a solution for the following queries:
1. Find the 10 countries with most projects
2. Find the top 10 major project themes (using column 'mjtheme_namecode')
3. Fill in missing values for major theme codes, and recalculate top 10 major project themes.

#### Let's begin by loading the dataset and investigating!

In [1]:
#Importing pandas, json, and numpy packages
import pandas as pd
import json
from pandas.io.json import json_normalize
from pandas.io.json import to_json
import numpy as np

In [2]:
#Read JSON data into DataFrame as well as a string format
json_df = pd.read_json('data/world_bank_projects.json')

with open('data/world_bank_projects.json') as json_file:
    json_data = json.load(json_file)

The world bank projects dataset contains observations on 500 projects with 50 variables recorded.  We can see a small section of the dataset below as well.  From the information provided, it appears that some columns contain missing values while others are complete.  As we dive into our queries, we will possibly need to investigate these missing values further.

In [3]:
json_df.shape

(500, 50)

In [4]:
json_df.head()

_id  approvalfy board_approval_month  \
0  {'$oid': '52b213b38594d8a2be17c780'}        1999             November   
1  {'$oid': '52b213b38594d8a2be17c781'}        2015             November   
2  {'$oid': '52b213b38594d8a2be17c782'}        2014             November   
3  {'$oid': '52b213b38594d8a2be17c783'}        2014              October   
4  {'$oid': '52b213b38594d8a2be17c784'}        2014              October   

      boardapprovaldate                                 borrower  \
0  2013-11-12T00:00:00Z  FEDERAL DEMOCRATIC REPUBLIC OF ETHIOPIA   
1  2013-11-04T00:00:00Z                    GOVERNMENT OF TUNISIA   
2  2013-11-01T00:00:00Z   MINISTRY OF FINANCE AND ECONOMIC DEVEL   
3  2013-10-31T00:00:00Z   MIN. OF PLANNING AND INT'L COOPERATION   
4  2013-10-31T00:00:00Z                      MINISTRY OF FINANCE   

            closingdate                              country_namecode  \
0  2018-07-07T00:00:00Z  Federal Democratic Republic of Ethiopia!$!ET   
1                   NaN                      Republic of Tunisia!$!TN   
2                   NaN                                   Tuvalu!$!TV   
3                   NaN                        Republic of Yemen!$!RY   
4  2019-04-30T00:00:00Z                       Kingdom of Lesotho!$!LS   

  countrycode                              countryname    countryshortname  \
0          ET  Federal Democratic Republic of Ethiopia            Ethiopia   
1          TN                      Republic of Tunisia             Tunisia   
2          TV                                   Tuvalu              Tuvalu   
3          RY                        Republic of Yemen  Yemen, Republic of   
4          LS                       Kingdom of Lesotho             Lesotho   

                         ...                           sectorcode source  \
0                        ...                          ET,BS,ES,EP   IBRD   
1                        ...                                BZ,BS   IBRD   
2                        ...                                   TI   IBRD   
3                        ...                                   JB   IBRD   
4                        ...                             FH,YW,YZ   IBRD   

   status  supplementprojectflg  \
0  Active                     N   
1  Active                     N   
2  Active                     Y   
3  Active                     N   
4  Active                     N   

                                              theme1  \
0      {'Percent': 100, 'Name': 'Education for all'}   
1  {'Percent': 30, 'Name': 'Other economic manage...   
2    {'Percent': 46, 'Name': 'Regional integration'}   
3  {'Percent': 50, 'Name': 'Participation and civ...   
4  {'Percent': 30, 'Name': 'Export development an...   

                                      theme_namecode    themecode   totalamt  \
0      [{'code': '65', 'name': 'Education for all'}]           65  130000000   
1  [{'code': '24', 'name': 'Other economic manage...        54,24          0   
2  [{'code': '47', 'name': 'Regional integration'...  52,81,25,47    6060000   
3  [{'code': '57', 'name': 'Participation and civ...        59,57          0   
4  [{'code': '45', 'name': 'Export development an...        41,45   13100000   

  totalcommamt                                                url  
0    130000000  http://www.worldbank.org/projects/P129828/ethi...  
1      4700000  http://www.worldbank.org/projects/P144674?lang=en  
2      6060000  http://www.worldbank.org/projects/P145310?lang=en  
3      1500000  http://www.worldbank.org/projects/P144665?lang=en  
4     13100000  http://www.worldbank.org/projects/P144933/seco...  

[5 rows x 50 columns]

In [5]:
json_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 50 columns):
_id                         500 non-null object
approvalfy                  500 non-null int64
board_approval_month        500 non-null object
boardapprovaldate           500 non-null object
borrower                    485 non-null object
closingdate                 370 non-null object
country_namecode            500 non-null object
countrycode                 500 non-null object
countryname                 500 non-null object
countryshortname            500 non-null object
docty                       446 non-null object
envassesmentcategorycode    430 non-null object
grantamt                    500 non-null int64
ibrdcommamt                 500 non-null int64
id                          500 non-null object
idacommamt                  500 non-null int64
impagency                   472 non-null object
lendinginstr                495 non-null object
lendinginstrtype            495 non

****
**Query #1**:  Find the 10 countries with the most projects.

Looking above at all of the column names, it appears that the column 'countryname' will be most useful for this situation.  Because we know there are 500 observations and the .info() method reports 500 non-null entries, it appears there are no missing or null values to clean in this column.

In [6]:
json_df.countryname.value_counts().head(10)

Republic of Indonesia              19
People's Republic of China         19
Socialist Republic of Vietnam      17
Republic of India                  16
Republic of Yemen                  13
Kingdom of Morocco                 12
Nepal                              12
People's Republic of Bangladesh    12
Republic of Mozambique             11
Africa                             11
Name: countryname, dtype: int64

****
**Query #2**: Find the top 10 major project themes (using column 'mjtheme_namecode').

Similar to before, it appears there are no missing or null values in this column as well.

In [7]:
json_normalize(json_data, 'mjtheme_namecode').name.value_counts().head(10)

Environment and natural resources management    223
Rural development                               202
Human development                               197
Public sector governance                        184
Social protection and risk management           158
Financial and private sector development        130
                                                122
Social dev/gender/inclusion                     119
Trade and integration                            72
Urban development                                47
Name: name, dtype: int64

Despite showing no null values, it DOES appear there are missing values (122 of them to be exact).  We will need to fill in these missing values with the appropriate theme before we can recalculate the top 10 major project themes.

****
**Query #3**: Fill in missing values for major theme codes, and recalculate top 10 major project themes.

In [8]:
#Assigning variable to the mjtheme_namecode column and print first 10 rows
theme = json_normalize(json_data, 'mjtheme_namecode')
theme.head(10)

code                                          name
0    8                             Human development
1   11                                              
2    1                           Economic management
3    6         Social protection and risk management
4    5                         Trade and integration
5    2                      Public sector governance
6   11  Environment and natural resources management
7    6         Social protection and risk management
8    7                   Social dev/gender/inclusion
9    7                   Social dev/gender/inclusion

My first step is to further investigate the mjtheme_namecode column in particular.  Looking at the first few rows, this column of the dataset has a code as well as a name for the project theme.  We can see that the name column has a missing entry, while code appears complete.  Before we continue, let's verify this is accurate.

In [9]:
#Checking to see if code column also has missing values
(theme.code == '').sum()

0

Great!  It appears there are no missing code values.  Next I will create a dictionary containing the unique pairing of code to name.

In [10]:
#Replace all empty strings '' with NaN values.  Then drop them, and convert to a dictionary where keys are code and values are names.
theme.set_index('code', inplace=True)
theme = theme.replace('', np.nan)
code_dict = theme.dropna().to_dict()
code_dict

{'name': {'8': 'Human development',
  '1': 'Economic management',
  '6': 'Social protection and risk management',
  '5': 'Trade and integration',
  '2': 'Public sector governance',
  '11': 'Environment and natural resources management',
  '7': 'Social dev/gender/inclusion',
  '4': 'Financial and private sector development',
  '10': 'Rural development',
  '9': 'Urban development',
  '3': 'Rule of law'}}

In [14]:
#Fill NaN values based on the code_dict created above.
theme = theme.fillna(code_dict)

#Requery the top 10 major project themes
theme.name.value_counts().head(10)

Environment and natural resources management    250
Rural development                               216
Human development                               210
Public sector governance                        199
Social protection and risk management           168
Financial and private sector development        146
Social dev/gender/inclusion                     130
Trade and integration                            77
Urban development                                50
Economic management                              38
Name: name, dtype: int64

And there are our top 10 major project themes after correcting the missing project names!

****
#### Miscellaneous
It wasn't originally part of the exercise, but I was curious if I could replace the missing code names within the JSON DataFrame itself.  Here is the code I wrote to complete that task, using the code_dict that I created earlier in these exercises.

In [16]:
#Isolate major theme column of JSON dataframe
mjtheme = json_df.mjtheme_namecode

#Checks each row in the DataFrame.  For each dictionary containg the code and name in that row, it reassigns the correct name based
#on the code_dict dictionary.
for proj in mjtheme:
    for theme in proj:
        theme['name'] = code_dict['name'][theme['code']]

#Reassigns the updated major project theme column to the JSON DataFrame.
json_df.mjtheme_namecode = mjtheme
